domain/network/state/subnet.go

```diff
@@ -0,0 +1,447 @@
+}
+
+const retrieveSubnetsStmt = `
+WITH fan_subnet AS (
+    SELECT * 
+    FROM subnet  
+        JOIN subnet_association  
+           ON subnet.uuid = subnet_association.associated_subnet_uuid
+           AND subnet_association.association_type_id = 0
+        WHERE subnet.subnet_type_id = 0
+    )
+SELECT     
+    subnet.uuid                          AS &Subnet.uuid,
+    subnet.cidr                          AS &Subnet.cidr,
+    subnet.vlan_tag                      AS &Subnet.vlan_tag,
+    subnet.space_uuid                    AS &Subnet.space_uuid,
+    provider_subnet.provider_id          AS &Subnet.provider_id,
+    provider_network.provider_network_id AS &Subnet.provider_network_id,
+    fan_subnet.cidr                      AS &Subnet.underlay_cidr,
+    availability_zone.name               AS &Subnet.az,
+    provider_space.provider_id           AS &Subnet.provider_space_uuid
+FROM subnet 
+    LEFT JOIN fan_subnet 
+    ON subnet.uuid = fan_subnet.subject_subnet_uuid
+    JOIN provider_subnet
+    ON subnet.uuid = provider_subnet.subnet_uuid
+    JOIN provider_network_subnet
+    ON subnet.uuid = provider_network_subnet.subnet_uuid
+    JOIN provider_network
+    ON provider_network_subnet.provider_network_uuid = provider_network.uuid
+    LEFT JOIN availability_zone_subnet
+    ON availability_zone_subnet.subnet_uuid = subnet.uuid
+    LEFT JOIN availability_zone
+    ON availability_zone_subnet.availability_zone_uuid = availability_zone.uuid
+    LEFT JOIN provider_space
+    ON subnet.space_uuid = provider_space.space_uuid`
+
+// GetAllSubnets returns all known subnets in the model.
+func (st *State) GetAllSubnets(
+	ctx context.Context,
+) (network.SubnetInfos, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space uuid condition to the query only if it's passed to the function.
+	q := retrieveSubnetsStmt + ";"
+
+	s, err := sqlair.Prepare(q, Subnet{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Subnets
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying subnets")
+	}
+
+	return rows.ToSubnetInfos(), nil
+}
+
+// GetSubnet returns the subnet by UUID.
+func (st *State) GetSubnet(
+	ctx context.Context,
+	uuid utils.UUID,
+) (*network.SubnetInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space uuid condition to the query only if it's passed to the function.
+	q := retrieveSubnetsStmt + " WHERE subnet.uuid = $M.id;"
+
+	s, err := sqlair.Prepare(q, Subnet{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Subnets
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{"id": uuid.String()}).GetAll(&rows))
```

Comment:
We should use `.Get`

Reply:
same answer as https://github.com/juju/juju/pull/16663#discussion_r1417343427